### ISSUE TITLE
New npm release please?

### ISSUE THREAD

**Issue by fiznool** [andrewplummer/Sugar#481]
Current version on npm is only 1.4.1. Would be great to get it up to 1.5.0. Thanks!

**Comment by andrewplummer** [comment]
hi... v1.5.0 actually hasn't been released yet... I've been very bad about working on this, but that stops today!

**Comment by vendethiel** [comment]
I'd really like to see this, as 1.4.1 was back in 17 Sep 2013 already...

**Comment by andrewplummer** [comment]
I know... :(

This has been a combination of lots of big changes plus trying to keep up with rolling changes and a healthy dose of procrastination thrown in. Rest assured it IS coming and I promise a lot of good stuff... Please just a little more patience!

**Comment by Globegitter** [comment]
Just wanted to say keep up the good work - looking forward to the release :)

**Comment by kobezzza** [comment]
:+1:

**Comment by andrewplummer** [comment]
Going to close this issue out as the next Sugar/npm release is full-steam ahead now. The next version will actually be 2.0.0. Tons of updates here, the most major is that prototype extension will be opt-in (first time mentioning this, will probably write a blog post about it -- long story short it's simply no longer workable to have extending happen by default when you may be depending on an intermediary library/plugin that wants to use Sugar).

What's changing:

- since extending is opt in this lets Sugar do npm exports properly instead of going in through the global object... sticky versioning issues with Sugar will no longer be a thing
- module.exports will expose the extending function itself, so if you still want prototype extension the difference will literally be two characters: `require('sugar')()`
- each Sugar package will now also be its own npm module (`sugar-array`, `sugar-string`, etc) which will all be maintained together and kept in sync.
- the `sugar` npm package will be the default build (basically everything but extra string methods and date locales)
- `sugar-core` will also be a module which defines no methods but will allow other plugin developers to create sugar plugins that end-users can then extend into natives as they see fit
- `sugar-full` will be a new module which includes every package in the Sugar repo
- `sugar-date-locales` will be a new module that is essentially the same as `sugar-date` but with all locales included.
- new build system is using gulp and much easier to use, making it super easy to make custom builds if the above isn't enough for you

Sugar on npm is about to get a lot better. Still a way off from release though so hang tight... will be soon!

**Comment by vendethiel** [comment]
> sugar-date-locales will be a new module that is essentially the same as sugar-date but with all locales included.


What's the size of them? I only need fr/en locales, and I've been considered trimming the other ones from my output js file.

**Comment by andrewplummer** [comment]
Depends but tend to be 1-3K when minified (French is closer to 1K, Russian is complex and closer to 3K).

I do expect a lot of need for requiring certain locales only... Actually now that I think about it maybe packaging all locales isn't a good idea as most people probably want to include them ad-hoc.

It's hard to know what to do here... the date locales are standalone in the browser as they are simply calling `Sugar.Date.addLocale`, however this global isn't exposed in npm, so it's tricky... what if we could do something like this:
```
var Sugar = require('sugar-dates');
require('sugar-dates/locales/fr')(Sugar);
```
The alternative would be:
```
var Sugar = require('sugar-dates');
var French = require('sugar-dates/locales/fr');
Sugar.Dates.addLocale('fr', French);
```

Not sure which is better, or if there is another solution, but open to suggestions...
Also if we can find a way not to build these locales into the main package, then maybe they could be included into the main `sugar` module as well...

**Comment by vendethiel** [comment]
I quite like the 2nd approach.

**Comment by andrewplummer** [comment]
I think I like it better as well... This will require a little tweaking as each locale will have to have it's own `exports`.

Also what about:
```
require('sugar-dates/locales/fr');
```
vs.
```
require('sugar-dates/fr');
```
?

If anyone else has other opinions please speak up!

**Comment by fiznool** [comment]
What about packaging each locale as a separate npm module? So if somebody
wanted French then you would

require('sugar-locale-fr')(Sugar)

And you could also build a convenience package for all locales if
necessary.

**Comment by andrewplummer** [comment]
So I would actually really like to discuss this point in detail. Is @fiznool's suggestion better?

What are the advantages and disadvantages of packaging multiple files in each release?
For example I've noticed that lo-dash is now modular as well and a number of different files are included in the same npm module. The official way is now:

```
require('lo-dash/array');
```

Conceivably, Sugar could do this as well and be just a single node module with a default build of `./sugar.js` but other packages also available (`require('sugar/array')` etc). Going this far "feels" wrong to me as the user would be downloading (quite a bit) more than they will actually use in the end. But given that Sugar now only has about 60K worth of date locales, maybe at least having these available in the `sugar-date` module would be simpler than having to download a separate module for each locale needed?

What do you think? Also I'm very interested in advantages/disadvantages about packaging in modules that I didn't mention here as I feel there's a lot of other points I'm not aware of...

**Comment by fiznool** [comment]
I guess on the node side it makes no difference, but for the browser, if
you don't need a locale then IMO it would be better to omit it from the js
that needs to be downloaded.

Separating out locales into separate modules would help here.

**Comment by andrewplummer** [comment]
Yeah I definitely agree about separating them, but there's a couple ways to do that. What if it was in the module but not the main package? In other words like this (`sugar-date` module files):

```
sugar-date.js
locales/fr.js
locales/es.js
locales/it.js
etc...
```

then in the `package.json`:

```
{
  "main": "sugar-date.js",
  "files": [
    "locales/fr.js",
    "locales/es.js",
    "locales/it.js",
    // etc... 
  ]
}
```

That way you can have everything in one module, but separated. The thing is that theres 2 "downloads"  -- *your* download on `npm install`, and the *user* download in the js you serve. It seems like if you have the locales split into separate files then it doesn't matter if those files are in the same npm module or their own (someone please correct me if I'm wrong about this).

**Comment by vendethiel** [comment]
packaging a different one per npm package would be a good thing, because then I'd be able to trim down the client-side package by quite a bit. Not sure how much effort that'd require from your side of things, though?

**Comment by andrewplummer** [comment]
Sorry to be pedantic but I want to be crystal clear because this seems to be confusing. You're saying you want to `trim down the client-side package`, which I totally understand, but this does not seem to be the same as with "packaging a different one [locale I'm assuming] per npm package".

Again, if all locales were included in the package **but not included in the main build**, then you should have access to them without being forced to build them together.

In other words you could do `npm install sugar-dates`, and would get all the locales, but in your code you choose which ones to use:

```
require('sugar');
require('sugar/locales/fr');
require('sugar/locales/it');
```

would result in a build with only the French and Italian locales in them. Unless I'm completely misunderstanding build systems (which I don't think I am), this should work.

**Comment by vendethiel** [comment]
I'm not sure my build system is actually going through the requires (which could be dynamic) to decide which file to include...

**Comment by andrewplummer** [comment]
What's your build system, and can you describe how you use it?

**Comment by andrewplummer** [comment]
I've just confirmed with Browserify and this seems to work very well. You can simply require the locales you need and browserify will build them into the final package. If you don't need them simply don't include. Will continue to have a look with a couple more build packages.

**Comment by vendethiel** [comment]
No, it's not bower, I'm using [commonjs-everywhere](https://github.com/michaelficarra/commonjs-everywhere). It should do that properly, however.
